activemodel/lib/active_model/attribute_registration.rb

```diff
@@ -10,17 +10,27 @@ module AttributeRegistration # :nodoc:
 
     module ClassMethods # :nodoc:
       def attribute(name, type = nil, default: (no_default = true), **options)
+        name = resolve_attribute_name(name)
         type = resolve_type_name(type, **options) if type.is_a?(Symbol)
 
-        pending = pending_attribute(name)
-        pending.type = type if type
-        pending.default = default unless no_default
+        pending_attribute_modifications << PendingType.new(name, type) if type || no_default
+        pending_attribute_modifications << PendingDefault.new(name, default) unless no_default
+
+        reset_default_attributes
+      end
+
+      def decorate_attributes(names = nil, &decorator) # :nodoc:
```

Comment:
@jonathanhefner We were using the block form of `attribute` at GitHub and I think it would be good to have a public method for this as it's a necessary utility unless we want users to guess at the DB schema's type.

Reply:
I agree it would be nice to have a public API for this.  Some design considerations:

* Should it target one or more attributes?  Or should it be able to target all attributes the way that `decorate_attributes` can?

* Should it accept a block or should it be more restricted?  For example, there was a previous proposal for something like:

    ```ruby
    class MyDecorator < ActiveModel::Type::Value
      def initialize(cast_type, **options)
        # ...
      end
    end

    ActiveModel::Type.register(:my_decorator, MyDecorator)
    # AND / OR
    ActiveRecord::Type.register(:my_decorator, MyDecorator)

    MyModel.decorate_attribute :my_attribute, :my_decorator, **options
    ```

* Naming: `decorate_attribute`?  `decorate_attribute_type`?  Simply `decorate`?  etc

In the mean time, if it's helpful, I can restore the block form of `attribute`.  Or you can monkey patch it back in with:

```ruby
module ActiveRecord::Attributes::ClassMethods
  def attribute(name, cast_type = nil, **options)
    super
    decorate_attributes([name]) { |name, type| yield type } if block_given?
  end
end
```